R/automation.R

```diff
@@ -128,45 +132,108 @@ setMethod("scriptSavepoint", "Script", function(x) {
+    # because the function is now broader, the argument is called `x`
+    # the following block is for backwards compatibility
+    args      <- as.list(sys.call())
+    args[[1]] <- NULL
+    if ('dataset' %in% names(args)){
+        warning(
+            'The argument `dataset` has been renamed to `x` because the function ',
+            'now supports more generic operations. ',
+            'While using `dataset` will continue to work, it will emit this warning.'
+        )
+        target <- list(x = args$dataset)
+        args$dataset <- NULL
+        dots$dataset <- NULL
+        do.call(
+            'runCrunchAutomation', c(target, args, dots),
+            quote = FALSE, envir = parent.frame()
+        )
+        return(invisible(NULL))
     }
-    if (length(script) != 1) script <- paste(script, collapse = "\n")
 
-    body <- list(body = script, ...)
+    filename <- NULL
+    if (is_file){
+        filename <- script
+        script   <- readLines(script, encoding = encoding, warn = FALSE)
+    }
+    script <- paste(script, collapse = '\n')
+    reset_automation_error_env()
+    automation_error_env$file                  <- filename
     automation_error_env$last_attempted_script <- script
 
+    sendCrunchAutomationRequest(x = x, script = script, ...)
+
+    # Provide feedback for dry_run success so that user is confident it was successful
+    if (isTRUE(dots$dry_run)) message("Script dry run was successful")
+
+    invisible(refresh(x))
+}
+
+setMethod("sendCrunchAutomationRequest", "CrunchDataset", function(
+        x,
+        script,
+        ...
+){
     crPOST(
-        shojiURL(dataset, "catalogs", "scripts"),
-        body = toJSON(wrapEntity(body = body)),
+        shojiURL(x, "catalogs", "scripts"),
+        body = toJSON(wrapEntity(body = list(body = script, ...))),
         status.handlers = list(`400` = crunchAutomationErrorHandler),
         progress.handler = crunchAutomationErrorHandler,
         config = add_headers(`Content-Type` = "application/json")
     )
-    # Provide feedback for dry_run success so that user is confident it was succesful
-    if (isTRUE(body$dry_run)) message("Script dry run was successful")
-    invisible(refresh(dataset))
-}
+    
+    invisible(NULL)
+})
+
+setMethod("sendCrunchAutomationRequest", "ProjectFolder", function(
+        x,
+        script,
+        is_file = string_is_file_like(script),
+        encoding = "UTF-8",
+        ...
+){
+    # project folders include a slot views with element execute,
+    # which gives us the URL to hit;
+    # but the account ('top-level folder', what you get from: `projects()`)
+    # is also of class ProjectFolder, but doesn't include this info;
+    # running CA scripts on the account is not supported currently
+    if (!is.crunchURL(x@views$execute)){
+        stop(
+            'not implemented yet'
+        )
+    }
+    
+    crPOST(
+        shojiURL(x, "views", "execute"),
+        body = toJSON(wrapView(value = script)),
```

Comment:
I'm not sure where the `...` should be, but ideally we'd be able to have `...` passed to the API somehow.

Reply:
I was basing this part on: https://help.crunch.io/hc/en-us/articles/10521594535309-Getting-started-with-folder-scripts

Where it says:

```
POST /api/projects/<project_id>/run/
{
  Element: shoji:view,
  Value: """script contents"""
}
```

So I'm also unsure where to put the extras. I'll investigate.

Reply:
Looking if scrunch does anything

Reply:
Based in conversation in automation channel, let's throw an error if `...` are not empty on a folder script.